File: backend/app/Models/ClassModel.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\Relations\HasMany;

class ClassModel extends Model
{
    protected $table = 'classes';

    protected $fillable = [
        'institute_id',
        'subject_id',
        'teacher_id',
        'name',
        'grade',
        'section',
        'room_number',
        'capacity',
        'day_of_week',
        'start_time',
        'end_time',
        'monthly_fee',
        'description',
        'is_active',
    ];

    protected $casts = [
        'capacity' => 'integer',
        'monthly_fee' => 'decimal:2',
        'is_active' => 'boolean',
    ];

    protected $appends = ['students_count', 'max_students'];

    public function getStudentsCountAttribute()
    {
        return $this->students()->count();
    }

    public function getMaxStudentsAttribute()
    {
        return $this->capacity;
    }

    public function institute(): BelongsTo
    {
        return $this->belongsTo(Institute::class);
    }

    public function subject(): BelongsTo
    {
        return $this->belongsTo(Subject::class);
    }

    public function teacher(): BelongsTo
    {
        return $this->belongsTo(Teacher::class);
    }

    public function students(): BelongsToMany
    {
        return $this->belongsToMany(Student::class, 'class_student', 'class_id', 'student_id')
            ->withPivot('enrolled_date');
    }

    public function attendances(): HasMany
    {
        return $this->hasMany(Attendance::class, 'class_id');
    }

    public function schedules(): HasMany
    {
        return $this->hasMany(Schedule::class, 'class_id');
    }

    public function grades(): HasMany
    {
        return $this->hasMany(Grade::class, 'class_id');
    }

    public function announcements(): HasMany
    {
        return $this->hasMany(Announcement::class, 'class_id');
    }

    public function scopeByGrade($query, string $grade)
    {
        return $query->where('grade', $grade);
    }

    public function isFull(): bool
    {
        return $this->students()->count() >= $this->capacity;
    }
}

File: admin-web/lib/models/class_model.dart
import 'package:json_annotation/json_annotation.dart';
import 'teacher.dart';
import 'student.dart';

part 'class_model.g.dart';

@JsonSerializable()
class ClassModel {
  final int id;
  final String name;
  final String? description;
  @JsonKey(name: 'subject_id')
  final int? subjectId;
  @JsonKey(name: 'teacher_id')
  final int? teacherId;
  final String? grade;
  final String? day;
  @JsonKey(name: 'start_time')
  final String? startTime;
  @JsonKey(name: 'end_time')
  final String? endTime;
  @JsonKey(name: 'monthly_fee')
  final double? monthlyFee;
  @JsonKey(name: 'max_students')
  final int? maxStudents;
  @JsonKey(name: 'is_active')
  final bool isActive;
  @JsonKey(name: 'created_at')
  final String? createdAt;
  @JsonKey(name: 'updated_at')
  final String? updatedAt;

  // Relationships
  final Teacher? teacher;
  final List<Student>? students;
  @JsonKey(name: 'students_count')
  final int? studentsCount;
  final Subject? subject;

  ClassModel({
    required this.id,
    required this.name,
    this.description,
    this.subjectId,
    this.teacherId,
    this.grade,
    this.day,
    this.startTime,
    this.endTime,
    this.monthlyFee,
    this.maxStudents,
    required this.isActive,
    this.createdAt,
    this.updatedAt,
    this.teacher,
    this.students,
    this.studentsCount,
    this.subject,
  });

  factory ClassModel.fromJson(Map<String, dynamic> json) =>
      _$ClassModelFromJson(json);

  Map<String, dynamic> toJson() => _$ClassModelToJson(this);
}

@JsonSerializable()
class Subject {
  final int id;
  final String name;
  final String? code;
  final String? description;
  @JsonKey(name: 'created_at')
  final String? createdAt;
  @JsonKey(name: 'updated_at')
  final String? updatedAt;

  Subject({
    required this.id,
    required this.name,
    this.code,
    this.description,
    this.createdAt,
    this.updatedAt,
  });

  factory Subject.fromJson(Map<String, dynamic> json) =>
      _$SubjectFromJson(json);

  Map<String, dynamic> toJson() => _$SubjectToJson(this);
}
